### PR TITLE
Allow readonly users to be Testers.

### PR DESCRIPTION
**Pull request by jgerigmeyer**
## Cute animal pic
_Because everyone likes pictures of [animals](https://unsplash.com/t/animals)._

## Link to Trello card (if applicable)
_If this PR relates to a Trello card, don't forget to reference this PR on the card as well._

- https://trello.com/c/6898xd2B

## Description
_The commit messages say what you did; this should explain why and/or how._
- [x] Description is in Trello card

## Steps to test/reproduce
_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._

- Log in as a readonly GH user
  - Assign yourself as a task Tester
  - You should be able to add/remove yourself, remove yourself as an Epic collaborator, and submit a Task PR review
- Log in as a read-write GH user
  - Add/remove a readonly user from a Task tester (not dev) assignment

## Show me
_Provide screenshots/animated gifs/videos if necessary._

Screen Shot 2021-05-12 at 3 10 19 PM

Screen Shot 2021-05-12 at 3 10 46 PM

Screen Shot 2021-05-12 at 3 10 55 PM

Screen Shot 2021-05-12 at 3 11 16 PM

### CODE REVIEW

File: metecho/api/jobs.py
Comment (jgerigmeyer) [inline metecho/api/jobs.py:719]: Readonly users cannot update the PR status, so we use the App account to do that, but the user account to add any review comments:

Screen Shot 2021-05-12 at 3 14 37 PM
Screen Shot 2021-05-12 at 3 14 48 PM

File: metecho/api/views.py
Comment (jgerigmeyer) [inline metecho/api/views.py:62]: @jerivas These are the two pieces where I'd like your help. Removing `enforce_push_permission=True` from the serializer for `assigned_qa` got me part of the way, but we also want readonly users to be able to make limited updates to tasks and epics:

- task: add/remove themselves (and only themselves) as `assigned_qa`
- epic: remove themselves from `github_users` (I suppose allowing adding is okay too, though I don't think we'll expose that via the UI)

**Comment by jgerigmeyer** [comment]
[As a user without write access, I want to help test tasks (acting as the Metecho application).](https://trello.com/c/6898xd2B/252-as-a-user-without-write-access-i-want-to-help-test-tasks-acting-as-the-metecho-application)

File: metecho/api/views.py
Comment (jerivas) [reply metecho/api/views.py:62]: What do you think of moving assignee/collaborator logic to a separate endpoint? Something like:

```http
POST or DELETE /api/tasks/<id>/assignees/

{
  "role": "qa",
  "user": "abc123",
}
```

And for Epics:

```http
POST or DELETE /api/epics/<id>/collaborators/

{
   "user": "abc123"
}
```

I propose this because I sense the permission checks will get pretty wild if we allow/deny updates based on the fields the user is trying to modify. Fully authorized users can use this new endpoint as well if we don't want the frontend to hit different URLs depending on permissions.

Comment (jerivas) [reply metecho/api/views.py:62]: Actually, we might be able to get away with it if we `PATCH` instead of `POST` on the regular endpoints, that should make it easier to detect if a user is updating **just** the `assigned_*` or `github_users` fields. That might be best if separate endpoints are too annoying to implement in the frontend.

Comment (jgerigmeyer) [reply metecho/api/views.py:62]: @jerivas I'm honestly fine with either approach. Both are relatively straightforward in the JS, so I'd defer to whichever you think will be cleaner for the API. (I'd lean toward all users using the same endpoint, though, and allowing the API to determine permissions.)

Comment (jerivas) [reply metecho/api/views.py:63]: Alright, turns out handling the assignees separately streamlined the vanilla `TaskSerializer` and required very minimal frontend changes. Everything works in my manual tests, just need to backfill tests and do something similar for collaborators

File: metecho/api/serializers.py
Comment (jgerigmeyer) [inline metecho/api/serializers.py]: @jerivas I think the order needs to be switched on these, so that in `_handle_reassign` we can compare the existing and new assignees before making the change.

```suggestion
        if "assigned_dev" in data:
            self._handle_reassign("dev", task, data, user, originating_user_id=user_id)
            task.assigned_dev = data["assigned_dev"]
        if "assigned_qa" in data:
            self._handle_reassign("qa", task, data, user, originating_user_id=user_id)
            task.assigned_qa = data["assigned_qa"]
```

Comment (jgerigmeyer) [reply metecho/api/serializers.py]: Fixed in https://github.com/oddbird/Metecho/commit/cadddfee10f9d33a1f554afeee8264ba3bb1b020

Comment (jerivas) [reply metecho/api/serializers.py]: Thanks, you hit this while writing the tests as well

File: metecho/api/views.py
Comment (jerivas) [reply metecho/api/views.py]: As of https://github.com/oddbird/Metecho/pull/1106/commits/85fb547b414aa14ada1c75c93516d0bd9e045bc4:

- Added `epic-collaborators` endpoint
- Added tests for both new endpoints

**Review by jerivas — APPROVED** [review]
LGTM 👍

**Review by jgerigmeyer — COMMENTED** [review]
🚀

**Pull request merged**
